CDAP-13349 add secure store integration to provisioners

albertshau opened the pull request:
Provisioner properties now support the $secure(key) macro function.
Lookups will be ignored, as will any other macro function.

Added some capabilities to the MacroParser to let it skip evaluation
of lookups, and to only try to evaluate certain macro functions.

Commit:
CDAP-13088 New MdsKey and storage for metadata

CuriousVini committed:
CDAP-13169 Adding transactional publishing and offset management for runtime monitor

Commit:
Metdata Plugin APIs

albertshau committed:
CDAP-13246 provisioner failure handling

Adding logic to handle failures during provisioning.

When a RetryableProvisionException is thrown, the method will
be retried up to a time limit. The time limit is hardcoded today,
but will be configurable per profile later.
Also added handling for scenarios where a cluster is request to
be created, but when polling for status, the cluster returns a
non-running status. In these scenarios, the cluster is usually
deleted, then the create is retried.

Also added logic to pick up tasks that were being executed while
CDAP was shut down. Each task will store state about what it is
about to do. If CDAP is shut down in the middle of a task, when
it comes back up, it will scan the state store and re-create
tasks that were in progress.

Also some cleanup to state transition logic by moving it out of
AppMetadataStore and into the respective ProgramRunStatus and
ProgramRunClusterStatus enums.

Fixing a bad test that relied on an invalid state transition, and
adding a test to make sure we can't go from pending to completed.

tonybach committed:
[CDAP-13079] Adds error handling in UI for MMDS
- Displays most errors in MMDS create view, list view and detail view using Alert component/banner
- Adds exclamation mark in case of model status when there's an error fetching the status
- Extracts current ActionCreator to ExperimentDetailActionCreator and ExperimentsListActionCreator
- Modifies Alert component to also take elements, and moves split failed message element to Alert banner
- Fixes experiment detailed page not scrolling

chtyim committed:
(CDAP-13252) Not to store private key in field

- Use Supplier instead

tonybach committed:
[CDAP-13461] Cloud/pipeline cleanup
- Fixes vertical alignment of namespace and cloud icon, and fixes double line for buttons in pipeline canvas
- Vertically aligns center tag text and + button
- Changes some labels to follow sentence-style heading
- Fixes vertical alignment of run profile in pipelines, and tags in tags modal
- Changes boldness of tag in spotlight modal to be consistent

ajainarayanan committed:
[MMDS][CDAP-13365][CDAP-13080] [CDAP-13458] Allows a user to resplit a previously failed split in a model
TL;DR-

 - Allows a user to replit a previously failed split from experiment detail view (Analytics -> Experiment 1 (Experiment detailed view)-> Model 3 (in Models table))
 - Allows the user to edit the directives from split step and edit the split from the algorithm selection step. This allows user to reiterate the changes in the same model instead of creating a new model everytime he/she decides to make a change.
 - Rounds off huge numbers to 4 digits after the decimal. Makes is readable in UI.
 - Fixes setting right schema while creating new models under an experiment.

Longer version -

  - Adds an active step to experiment create store to switch between different stages (Dataprep <-> Data Split <-> Algorithmselection)
  - Adds a enum(sortof) for ModelStatuses to reflect backend
  - Adds api to update directives in model + delete (unassign) a split in model
  - Updates Experiment detailed view to use newly create ModelStatus enum
  - Adds a generic function to round decimal numbers in experiments
  - Updates Split step to use directives and schema from the model instead of from the experiment

Commit:
(CDAP-13409) Change SDK default version to spark 2

albertshau committed:
CDAP-13480 fix test to wait for service status before getting url

Fix a test to wait for the service to be running before getting
the service URL.

dereklwood committed:
detect hdp version and invoke version-specific ambari interaction functions

albertshau committed:
CDAP-13486 fix waitForStopped

The method was incorrectly waiting for the state to not be
RUNNING instead of waiting for it to be STOPPED. This is incorrect,
as waitForStopped will return if it sees the state as STARTING.

ajainarayanan committed:
Adds displayName to avoid js minification to mangle component's display name

albertshau committed:
CDAP-13405 increase default twill reserved memory

Increasing the default twill reserved memory from 300 to 768.
I have seen YARN kill containers due to physical memory limits
when this is set to 512, but not when it is set to 768.

albertshau committed:
CDAP-13489 fix flaky dynamic partitioning test

This test became flaky when we introduced the PENDING program
run state. Replaced a more complicated and brittle way of waiting
until the program was done with a call to waitForStopped.

albertshau committed:
CDAP-13480 wait for service running before getting url

Fixing another place where we were not waiting for the service
to be running before getting the service URL.

albertshau committed:
CDAP-13491 fix mr resources test

Fix a test for overridding system resources that broke after
updating twill reserved memory. Modifying the test so that it is
immune to updates to these default values.

albertshau committed:
CDAP-13493 remove hadoop operational stats

Removing YARN, HDFS, and HBase operational stats. These turned
out to not be very useful for people, while being actively
confusing in certain cluster setups. For example, with a secure,
HA namenode, the HDFS operational stats would log warnings that
confused most users.

In addition, these stats will be confusing once we add the cloud
integrations planned for 5.0.0

chtyim committed:
(CDAP-13245) Remove all web-app related logic

albertshau committed:
CDAP-13489 re-fix dynamic partitioning test

The previous fix was not correct, as it was too aggressive in
removing logic. The call to start a program does not guarantee
any state change by the time the call returns. We therefore cannot
just call waitForStopped() right after calling start().

We really need to change the way a program run is handled in
unit tests.

albertshau committed:
CDAP-13501 fix flaky run history tests

Fix a race condition where a call to list run records is made first,
then calls to fetch individual run records are made and the results
compared. This is subject to race conditions where the cluster state
has changed. For example, when listing, a run record might be in the
DEPROVISIONING state. When it is individually fetched, it could be
in the DEPROVISIONED state. Instead of comparing entire run records,
just comparing run record program state, since that is really all that
is important for the test.

Also performing cleanup to remove duplicate code and to remove a bunch
of warnings.

elfenheart committed:
collapse status STOPPED to STOPPED and KILLED

albertshau committed:
CDAP-13489 another dynamic partioning test fix

The previous fix can fail due to a race in program status
(CDAP-13296) where run records are not fetched in a single
transaction. If a run transitions state during a status call,
the status can incorrectly be returned as STOPPED.
Instead, just wait for the number of failed runs to increase,
which should always indicate that the run has completed.

albertshau committed:
CDAP-13493 remove unneeded hadoop excludes from pom

ajainarayanan committed:
[CDAP-13455][CDAP-13515] Fixes choosing profile for schedule + fixes the "save" button behavior in pipeline configure step

 - Adds a provisioner store to be reused across all profile references
 - Fixes Profiles DetailedView to use Provisioners Store
 - Fixes Profiles ListView to use Provisioners Store
 - Fixes ProfilesListView in pipelines to use Provisioner Store:
 - Fixes PipelineScheduler to use Provisioner Store + to show the right provisoner label for a selected profile
 - Removes the unused ProfilesListView
 - Fixes AbstractWidget's StateWrapper to allow accepting a function for value to get upto date value on each re-render
 - Modifies ProfileCustomization popover to save proper values on open/close accordion panes
 - Fixes pipeline scheduler to close profiles dropdown on select
 - Removes saving profiles to schedule from profile customize popover
 - Removes the restriction of saving anything in pipeline configuration modeless
 - Fixes getProvisionerMap api to be more concrete to remove unnecessary code duplication during its usage
 - Removes check for empty run time arguments before saving it to preference store
 - Fixed onhover styling for profiles table view

albertshau committed:
CDAP-13296 fix race in program status call

The call to get a program status was suspectible to races because
run records for the program were not read in a single transaction.
If run state happened to change in between run record reads, the
state could incorrectly be returned as stopped.

This race condition would cascade and cause other tests to fail
that would wait for a program run to complete. If one of those
tests ran into this scenario, the test would move on assuming
the program had stopped when it in fact was still running.

Fixed to read run records in a single transaction, and added
missing unit tests.

Also fixing a bug where a suspended program would have a stopped
status instead of a running status.

chtyim committed:
(CDAP-13252) Added HttpsEnabler helper class

- Helps enabling HTTPS in various server and client usage context

yaojiefeng committed:
CDAP-13354 fix incorrect response code for putting a profile

yaojiefeng committed:
CDAP-13512 Allow updating a profile if it already exists

sagarkapare committed:
CDAP-13269 REST api with mock data and relavant classes in the proto.

yaojiefeng committed:
CDAP-13494 adding enable/disable functionality

elfenheart committed:
add routing to field level lineage page

elfenheart committed:
add routing to field level lineage page

chtyim committed:
(CDAP-13252) Rename SSHKeyInfo to SecureKeyInfo

- Allow the class to store all secure key related information instead of just SSH

tonybach committed:
[CDAP-13457][CDAP-13470][CDAP-13476] Bugfixes/improvements in Reports UI
- Fixes Popover component closing on every click
- Changes how we display runtime arguments in reports detail page
- Still reverts back to previous statuses when user clicks outside, but displays the statuses as user selects them, and changes Apply link to button
- Makes runtime args column fixed width and displays in one line, and adds title attributes to all table headers and cells
- Fixes so that gear popover on a report shows up correctly
- Don't show effect when hovering over grid row in reports detailed table

ajainarayanan committed:
Updates the profile property lock/unlock tooltips text

Commit:
files moved to cdap-program-report-base and cdap-program-report2_2.11

Commit:
packaging for cdap-master and cdap-standalone

Commit:
use variable for spark version dir

ajainarayanan committed:
[MMDS][Cloud][UI] Fix language in MMDS and Profiles in pipelines

 - Minor language change to profile create view
 - Fix Dataproc provisioner name & update maven module name
 - Modifies title for create experiment/model vs add model to an existing experiment
 - Adds provisioner label dynamically based on selected provisioner during profile creation

sagarkapare committed:
CDAP-13269 Javadoc for the proto classes for Field Level Lineage.

Commit:
[CDAP-13305] Program report app spark1 compatibility

Commit:
fix poms

chtyim committed:
(CDAP-13252) Enable two-ways HTTPS authentication for runtime monitor

ajainarayanan committed:
Fixes text changes in split step in MMDS

tonybach committed:
[CDAP-13474][CDAP-13475] Adds i18n for Ops Dashboard and Reports
- Adds i18n for Ops Dashboard
- Adds i18n for Reports
- Changes time to 12-hour format, since we're showing AM/PM

ajainarayanan committed:
Adds anchor tags to have underline during hover to give a visual feedback for the user

elfenheart committed:
add field level lineage summary popover

ajainarayanan committed:
[MMDS][CDAP-13128] UX fixes in experiment detailed view in MMDS

 - Adds an option to specify tooltip text for CopyableId if the test to copy is too long
 - Adds a Copy to Clipboard for the directives in model
 - Adds the ability to hide tooltip on hover for copyableid component to show only the copied message:
 - Adds features count in features popover under model
 - Removes all references of "Deployed" info from experiments
 - Adds color range for experiment metrics bar chart in experiments detailed view
 - Adds color scheme for variable model status for pie chart in experiment detailed view
 - Modifies variables.scss to export colors to be used in js world
 - Modifies IconSVG to allow to accept style tag for cases when required in js world
 - Adds an alias for styles in cdap webpack config

elfenheart committed:
add fll search capability

Commit:
CDAP-13517-Metrics integration

sagarkapare committed:
CDAP-13267 Added ability to get the fields associated with the destination endpoints and summary in both incoming and outgoing direction.

chtyim committed:
(CDAP-13532) Fix the workflow resources setting to honor runtime arguments

tonybach committed:
[CDAP-13520] Reports detail bar chart changes
- Adds select element to show legends based on either run status or start method
- Changes graph based on status or start method option, and removes legends reducer from DashboardStore since we don't need it anymore
- Highlights selected bucket in chart
- Adds additional layer on top to handle clicks, instead of attaching click handlers to bars, and renames some classes
- Selects last bucket by default instead of first bucket, and addresses other code comments

ajainarayanan committed:
[MMDS][CDAP-13385] Improve representation of model metrics

 - Adds an aditional option to format decimal numbers to humanReadableNumber utility function
 - Formats large numbers shown as part of model metrics in experiment detailed view

ajainarayanan committed:
[MMDS][CDAP-13383] Fix directives representation for a model consistent

 - Shows hint of directives applied to a model. Giving it similar treatment to features under a model

anwar6953 committed:
Fix DataProcClient NPE

ajainarayanan committed:
[MMDS][CDAP-13547] Fix button label while creating experiment vs adding a model in MMDS

 - Switch UnControlledPopover with Popover for addmodel/create experiment popover
 - Fixes the label for add model btn to show appropriate label while creating experiment vs adding a model
 - Fixes the title of the filebrowser while choosing a file to experiment with
 - Fixes setting active step while creating an experiment
 - Automatically open the popover when the user chooses a file and lands in dataprep UI

elfenheart committed:
add operations modal

tonybach committed:
[CDAP-13471][CDAP-13520] Adds legends to runs chart, and other improvements in Reports
- Makes runs list grid sortable, and removes highlight on hovering over a row, since it's not clickable
- Adds start time column to runs table, and uses it as the default column to sort by
- Adds legends on x-axis and left-right y axes
- Removes the runs list toggle, to always show runs list
- Converts values to more user-friendly format, and changes last 24 hours btn color
- Adds moment-timezone library to be able to show time zone on x-axis legend, and addresses code comments
- Changes type selector checkbox color to blue, and adds colored circles next to statuses in runs table
- Adds last updated label
- Adds 'across # namespaces' label
- Adds vertical separator between two dates
- Disables last 24 hours once clicked on
- Fixes sorting by status

Commit:
Fix 'rename-program-report-jars' in cdap-master pom

ajainarayanan committed:
Asd'

[CDAP-13536] CDAP UI Header and Footer Changes

 - Fixes footer to change copyright information
 - Removes 'Product Website' and 'Support' links from CDAP dropdown in header

elfenheart committed:
add basic time range picker

rohitsinha54 committed:
CDAP-13088 Support backward compatibility for Metadata APIs

chtyim committed:
(CDAP-13342) Process metrics from TMS on SDK

albertshau committed:
CDAP-13349 add secure store integration to provisioners

Provisioner properties now support the $secure(key) macro function.
Lookups will be ignored, as will any other macro function.

Added some capabilities to the MacroParser to let it skip evaluation
of lookups, and to only try to evaluate certain macro functions.

anwar6953 committed:
Update DataProcClient.java

anwar6953 committed:
Address comment

albertshau committed:
CDAP-13573 configure sandbox with secure storage

Include configuration in the sandbox such that the secure store
APIs will be available by default. Though secure storage is not
really a concern in local development environments, users may
still want to try out secure macros in the sandbox.

ajainarayanan committed:
Updates yarn.lock file for moment-timezone

tonybach committed:
[CDAP-13562] Implements profile enabling/disabling in UI
- Shows enabled/disabled status in profiles list view table
- Adds status toggle widget in profile detail view using ToggleSwitch component, and modifies it to accept custom on/off labels
- Adds enabled/disabled status column to profiles list view in pipelines, greys out disabled profile, and automatically switches to default when previously selected profile is disabled
- Don't show enable/disable profile switch for the system 'default' profile
- Shows profile status column in profile preview
- Don't change selection to default in the profiles list view is disabled, and sets default profile selection to 'default' instead of showing 'Select a Profile'
- Fixes the styles for angular switch directive to be within the component
- Shows confirmation modal when disabling a profile, and checks profileIsEnabled inside onProfileSelectHandler
- Extracts profile status toggle component

yaojiefeng committed:
CDAP-13432 publish event to tms to update profile metadata

albertshau committed:
CDAP-13391 log provisioner messages in program context

Modify the provisioner tasks to use a program logging context
instead of a system logging context in order to tie the logs to
the program run and not the system logs. Even though the provisioner
runs in the CDAP master, it's logs belong with the program run,
otherwise a provisioning error will cause a program run to fail
without any feedback given to the user.

Also adding a cluster end time to each run record. This is needed
because the log handler restricts the log query end time using
the end time in the run record if it exists. However, the end time
in the run record is the program run end, whereas the cluster end
time is usually minutes later. Adjusted the log handler to restrict
using the cluster end time, so that the deprovisioning logs will
be included in the program logs.

Removing the expiresAt field from the cluster proto since it is
currently not used. It will be re-introduced when we implement
the feature that allows keeping a cluster around after stop or
failure of a program run.

albertshau committed:
CDAP-13586 set secure store provider for clusters

Set the secure store provider to 'none' for clusters. This is
to prevent errors when it is configured to be something else on
the CDAP master. In the future, this will be configurable based
on the profile.

albertshau commented:
https://builds.cask.co/browse/CDAP-DUT6494

tonybach committed:
[CDAP-13473] Adds pagination to Reports detail view in UI
- Adds pagination to reports detail view
- Modifies PaginationWithTitle to not show the pagination component if there's only 1 page

gsps1 committed:
CDAP-13352 adding heart beat thread to Program State writer

gsps1 commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/MacroParser.java:
```diff
@@ -72,31 +79,40 @@ private String parse(String str, int depth) throws InvalidMacroException {
     }
 
-    MacroMetadata macroPosition = findRightmostMacro(str);
+    MacroMetadata macroPosition = findRightmostMacro(str, str.length());
```
hmm, can you add some comment/clarification, the api documentation for `findRightmostMacro` says 2nd param is the position to start the search for macro from, if thats the string length, this will always be empty or null right ?

albertshau replied:
ok i can clarify it in the comments. It's the position to start searching backwards from, similar to String's .lastIndexOf() method.

albertshau committed:
adding more comments

ajainarayanan committed:
Updates defaults for master and worker node configurations for dataproc provisioner for profile creation

elfenheart committed:
add link to field level lineage

albertshau committed:
CDAP-13358 rename 'default' profile to 'native'

This rename is to avoid confusion, since the word default has
meaning about how a profile is used, but not what the profile is.
For example, the default profile for a namespace is actually
whatever profile is set for that namespace, which might be a user
created profile. Changing it to 'native' so that it describes
what the profile does.

Commit:
[CDAP-13263] Consume Metadata Changes from TMS

rohitsinha54 committed:
CDAP-13088 Make metadata version independent

sagarkapare committed:
CDAP-13601 Added rules for field lineage endpoints in RouterPathLookup.

ajainarayanan committed:
Fixes bower registry to be the new one as bower is deprecating its old registry

Commit:
[CDAP-13609] fix flaky MessagingMetricsProcessorServiceTest

sagarkapare committed:
CDAP-13593 No need to start/stop SystemArtifactLoader service when running preview.

albertshau committed:
CDAP-13614 avoid a guaranteed NPE when task state is not found

We don't expect the task state to be missing, but will fix that
in a follow up PR.

ajainarayanan committed:
Adds corresponding UI change to choose native profile when no profile is chosen in UI

albertshau committed:
CDAP-13521 fix profile property override

Fixed a bug where profile properties provided at runtime were not
correctly overriding whatever currently exists in the profile.
In addition, properties not set in the profile were being ignored
completely.

Added a method to ProfileService to do the property overrides and
added a unit test for it.

elfenheart committed:
UI fix non existing status field for future scheduled run

albertshau committed:
CDAP-13338 restrict profile program types

Make sure only workflows look up the profile from the runtime
args. All other program types will run using the native profile.

albertshau committed:
CDAP-13623 fix flaky service test

Fix the test to wait for the service to be running before checking
that it is available.

albertshau commented:
messed up the rebase
